### ISSUE TITLE
cupyx.scipy.sparse.linalg.eigsh gives wrong and inconsistent results when solving large k and using large ncv. Loss of orthogonality present..

### ISSUE THREAD

**Issue by RisaKirisu** [cupy/cupy#7355]
### Description

Pretty much the title. For the same sparse hamiltonian matrix produced by the code below , the result with ```k=50``` is consistant and matches the result from ```scipy.sparse.linalg.eigsh```, but if I try ```k=300```, the result would be very inconsistent between each run and also way off. Using 64-bit arithmatic helps to a point, but doesn't solve the problem of inconsistency. Reducing ```ncv``` also help. On the contrary, even using 32-bit float point,  ```scipy.sparse.linalg.eigsh``` would hold correct and consistent.

### To Reproduce

```py
import numpy as np
from cupyx.scipy.sparse import kron
import cupy as cp
import cupyx
import scipy.sparse.linalg as la

def GetHermitian(N, S, delta_o, precision=cp.float64):

    def kron3(A,B,C):
        D= kron(A,kron(B,C))
        return D

    sz=cp.array([[1.0,0],[0,-1]])
    sx=cp.array([[0,1.0],[1,0]])
    sy=cp.array([[0,1.0j],[-1j,0]])

    if precision == cp.float64:
        cprecision = cp.cdouble
    else:
        cprecision = cp.csingle

    sz = cupyx.scipy.sparse.csr_matrix(sz, dtype=precision)
    sx = cupyx.scipy.sparse.csr_matrix(sx, dtype=precision)
    sy = cupyx.scipy.sparse.csr_matrix(sy, dtype=cprecision)

    t0 = np.eye(N)
    t1 = np.diag(np.ones(N - 1),k=1) + np.diag(np.ones(N - 1),k=-1)
    t1[0,-1]=1
    t1[-1,0]=1
    tt1 = np.diag(np.ones(N - 2),k=2) + np.diag(np.ones(N - 2),k=-2)
    tt1[0,-2]=1
    tt1[-2,0]=1
    tt1[1,-1]=1
    tt1[-1,1]=1

    ###############
    t =0
    tt = 1
    ttt = 0
    tttt=0
    mu = 0

    t0 = cp.array(t0, dtype=precision)
    t1 = cp.array(t1, dtype=precision)
    tt1 = cp.array(tt1, dtype=precision)

    t0 = cupyx.scipy.sparse.csr_matrix(t0)
    t1 = cupyx.scipy.sparse.csr_matrix(t1)
    tt1 = cupyx.scipy.sparse.csr_matrix(tt1)


    t = 0.25
    tt = -0.031863
    ttt = 0.016487
    tttt = 0.0076112
    mu = -0.16235
    ###############
    T = -mu*kron3(sz,t0,t0)\
    -t*(kron3(sz,t1,t0)+kron3(sz,t0,t1))\
    -tt*kron3(sz,t1,t1)-ttt*(kron3(sz,tt1,t0)+kron3(sz,t0,tt1))\
    -tttt*kron3(sz,tt1,tt1)

    ###########################

    dform = 1/2*kron(t1,t0)- kron(t0,t1)

    S = delta_o * cupyx.scipy.sparse.diags(S.flatten(), dtype=cprecision)
    S = S @ dform + dform @ S
    S = cupyx.scipy.sparse.vstack(
        [cupyx.scipy.sparse.hstack([cupyx.scipy.sparse.csr_matrix((N**2, N**2)), S]),
        cupyx.scipy.sparse.hstack([S.getH(), cupyx.scipy.sparse.csr_matrix((N**2, N**2))])],
        dtype=S.dtype
        )
    T += S
    print(T.dtype)
    print("Hermitian Generated. Shape: " + str(T.shape))
    return cupyx.scipy.sparse.csr_matrix(T)

# Construct the matrix
N = 70
delta_o = 0.06
S = np.ones((N, N))
T = GetHermitian(N, S, delta_o, precision=np.float32)
# T.shape: (9800, 9800), T.nnz: 205800

E, V = cupyx.scipy.sparse.linalg.eigsh(T, k=300, which='LA')
Ecpu, Vcpu = la.eigsh(T, k=300, which='LA')    # For reference
```

### Installation

Wheel (`pip install cupy-***`)

### Environment

```
OS                           : Windows-10-10.0.22621-SP0
Python Version               : 3.9.13
CuPy Version                 : 11.5.0
CuPy Platform                : NVIDIA CUDA
NumPy Version                : 1.24.1
SciPy Version                : 1.10.0
Cython Build Version         : 0.29.33
Cython Runtime Version       : None
CUDA Root                    : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8
nvcc PATH                    : C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.8\bin\nvcc.EXE
CUDA Build Version           : 11080
CUDA Driver Version          : 12000
CUDA Runtime Version         : 11080
cuBLAS Version               : (available)
cuFFT Version                : 10900
cuRAND Version               : 10300
cuSOLVER Version             : (11, 4, 1)
cuSPARSE Version             : (available)
NVRTC Version                : (11, 8)
Thrust Version               : 101501
CUB Build Version            : 101501
Jitify Build Version         : 4a37de0
cuDNN Build Version          : (not loaded; try `import cupy.cuda.cudnn` first)
cuDNN Version                : (not loaded; try `import cupy.cuda.cudnn` first)
NCCL Build Version           : None
NCCL Runtime Version         : None
cuTENSOR Version             : None
cuSPARSELt Build Version     : None
Device 0 Name                : NVIDIA GeForce RTX 2080 Ti
Device 0 Compute Capability  : 75
Device 0 PCI Bus ID          : 0000:07:00.0
```

**Comment by RisaKirisu** [comment]
I had a few suspicion, one of which is that with larger number of ```ncv``` due to large ```k```, the basis in Cupy's implementation of Thick-restart lanczos iteration lost orthogonality quickly and thus lead to wrong and inconsistent results. I tested this by adding a few lines into Cupy's source code that calculate the orthogonality of the basis between each restart.

```py
def eigsh(a, k=6, *, which='LM', ncv=None, maxiter=None, tol=0, return_eigenvectors=True):
  # Initial setup
  ...
  V = cupy.empty((ncv, n), dtype=a.dtype)   # This is the set of basis
  # Set initial vector
  ...
  # Choose Lanczos implementation, unconditionally use 'fast' for now
  ...
  # Lanczos iteration
  ...
  # Compute residual
  ...
  # Track orthogonality of the basis
  I = cupy.eye(ncv, dtype=a.dtype)
  ortho = []

  while res > tol and iter < maxiter:
    # Track orthogonality of the basis
    ortho.append(cupy.linalg.norm(V @ V.conj().T - I, ord='fro'))
    # Rest of restart
    ...

  # Return of results and cupy.array(ortho)
```

Level of orthogonality is here measured by Frobenius_Norm(V* @ V - I)
This confirmed my suspicion. As you can see below, in the ```k=50``` case (top), the basis remained fairly orthogonal, while in the ```k=300``` case (bottom), the basis loses orthogonality real fast.
![image](https://user-images.githubusercontent.com/28692567/215268378-b1827986-022e-4dac-bb28-cfe2643dcb84.png)
![image](https://user-images.githubusercontent.com/28692567/215268381-c4ac6860-ea3f-49c7-ba89-6d5b03caf5b4.png)

**Comment by emcastillo** [comment]
Thanks a lot for checking into this!
We have been aware of this problem for some time now, but we haven't had time to debug it yet. The info you provide here will be very valuable to fix the issue! Thanks a lot again, and if you feel in the mood, maybe you can help us to fix this 😅
cc @anaruse

**Comment by RisaKirisu** [comment]
After finding out the reason, it's actually quite easy to spot the minor implementation bug of the thick-restart algorithm. I fixed the orthogonalization part. The full orthogonalization in the current implementation is more stable than the orthogonalization step described in the paper, but apperently not enough for large ```k``` and ```ncv```. Doing another iteration of full orthogonalization is expensive as k get large, although effective. So I used the orthogonalization step in the paper plus an iteration of full reorthogonalization. Judging by the level of orthogonality of output eigen vectors, result of the fixed version is even cleaner than Scipy's result.

Below is the orthogonality level of a ```k=1000``` run
![image](https://user-images.githubusercontent.com/28692567/215458218-53e1125f-b031-4c27-bc70-c7d5d2fd548a.png)
